fdbserver/worker.actor.cpp

```diff
@@ -1356,20 +1392,77 @@ UpdateWorkerHealthRequest doPeerHealthCheck(const WorkerInterface& interf,
+	state Transaction tr(db);
+	try {
+		tr.setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
+		tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+		tr.setOption(FDBTransactionOptions::LOCK_AWARE);
+		std::vector<std::pair<StorageServerInterface, ProcessClass>> results =
+		    wait(NativeAPI::getServerListAndProcessClasses(&tr));
+		PrimaryAndRemoteAddresses storageServers;
+		for (auto& [ssi, _] : results) {
+			const auto primaryDCId = getPrimaryDCId(dbInfo->get());
+			const bool primarySS = ssi.locality.dcId().present() && primaryDCId.present() &&
+			                       ssi.locality.dcId().get() == primaryDCId.get();
+			const bool remoteSS = ssi.locality.dcId().present() && primaryDCId.present() &&
+			                      ssi.locality.dcId().get() != primaryDCId.get();
+			if (SERVER_KNOBS->GRAY_FAILURE_ALLOW_PRIMARY_SS_TO_COMPLAIN && primarySS) {
+				storageServers.primary.push_back(ssi.address());
+				if (ssi.secondaryAddress().present()) {
+					storageServers.primary.push_back(ssi.secondaryAddress().get());
+				}
+			} else if (SERVER_KNOBS->GRAY_FAILURE_ALLOW_REMOTE_SS_TO_COMPLAIN && remoteSS) {
+				storageServers.remote.push_back(ssi.address());
+				if (ssi.secondaryAddress().present()) {
+					storageServers.remote.push_back(ssi.secondaryAddress().get());
+				}
+			}
+		}
+		ret = storageServers;
+	} catch (Error& e) {
+		if (e.code() != error_code_actor_cancelled) {
+			TraceEvent("GetStorageServersError").error(e);
+		}
+	}
+	return ret;
+}
+
 // The actor that actively monitors the health of local and peer servers, and reports anomaly to the cluster controller.
 ACTOR Future<Void> healthMonitor(Reference<AsyncVar<Optional<ClusterControllerFullInterface>> const> ccInterface,
                                  WorkerInterface interf,
                                  LocalityData locality,
                                  Reference<AsyncVar<ServerDBInfo> const> dbInfo,
                                  Reference<AsyncVar<bool>> enablePrimaryTxnSystemHealthCheck) {
 	state UpdateWorkerHealthRequest req;
+	state Optional<Database> db;
+	if (SERVER_KNOBS->GRAY_FAILURE_ALLOW_PRIMARY_SS_TO_COMPLAIN ||
+	    SERVER_KNOBS->GRAY_FAILURE_ALLOW_REMOTE_SS_TO_COMPLAIN) {
+		db = openDBOnServer(dbInfo, TaskPriority::DefaultEndpoint, LockAware::True);
+	}
 	loop {
-		Future<Void> nextHealthCheckDelay = Never();
+		state Future<Void> nextHealthCheckDelay = Never();
 		if ((dbInfo->get().recoveryState >= RecoveryState::ACCEPTING_COMMITS ||
 		     enablePrimaryTxnSystemHealthCheck->get()) &&
 		    ccInterface->get().present()) {
 			nextHealthCheckDelay = delay(SERVER_KNOBS->WORKER_HEALTH_MONITOR_INTERVAL);
-			req = doPeerHealthCheck(interf, locality, dbInfo, req, enablePrimaryTxnSystemHealthCheck);
+			state Optional<PrimaryAndRemoteAddresses> storageServers;
+			if (db.present()) {
+				Optional<PrimaryAndRemoteAddresses> storageServers_ = wait(getStorageServers(db.get(), dbInfo));
```

Comment:
can be simplified as `store(storageServers, getStorageServers(db.get(), dbInfo));

Reply:
Reads better, thanks for telling about this utility. Fixed in #11757. Since `store` returns `Future`, also added `wait` to your code suggestion.